File: android/app/src/main/java/com/collectrndemo/modules/collect/field/core/BaseCollectView.java
package com.collectrndemo.modules.collect.field.core;

import android.content.Context;
import android.util.TypedValue;

import androidx.annotation.NonNull;

import com.collectrndemo.modules.collect.VGSCollectOnCreateViewInstanceListener;
import com.collectrndemo.modules.util.ResourceUtil;
import com.facebook.react.uimanager.ThemedReactContext;
import com.facebook.react.uimanager.ViewGroupManager;
import com.facebook.react.uimanager.annotations.ReactProp;
import com.verygoodsecurity.vgscollect.view.InputFieldView;
import com.verygoodsecurity.vgscollect.widget.VGSTextInputLayout;

public abstract class BaseCollectView<T extends InputFieldView> extends ViewGroupManager<VGSTextInputLayout> {

    private static final int DEFAULT_CORNER_RADIUS = 8;

    private final VGSCollectOnCreateViewInstanceListener listener;

    protected VGSTextInputLayout layout;

    protected T input;

    public BaseCollectView(VGSCollectOnCreateViewInstanceListener listener) {
        this.listener = listener;
    }

    protected abstract String getHint();

    protected abstract T createInput(ThemedReactContext reactContext);

    @NonNull
    @Override
    protected VGSTextInputLayout createViewInstance(@NonNull ThemedReactContext reactContext) {
        layout = createLayout(reactContext);
        input = createInput(reactContext);
        layout.addView(input);
        listener.onCreateViewInstance(input);
        return layout;
    }

    @ReactProp(name = "fiendName")
    public void setFieldName(VGSTextInputLayout view, String text) {
        input.setFieldName(text);
    }

    @ReactProp(name = "fontSize")
    public void setFontSize(VGSTextInputLayout view, int size) {
        input.setTextSize(TypedValue.COMPLEX_UNIT_DIP, size);
    }

    @ReactProp(name = "padding")
    public void setPadding(VGSTextInputLayout view, int padding) {
        int paddingDp = ResourceUtil.convertPxToDp(view.getContext(), padding);
        input.setPadding(paddingDp, paddingDp, paddingDp, paddingDp);
    }

    @ReactProp(name = "hint")
    public void setHint(VGSTextInputLayout view, String text) {
        view.setHint(text);
    }

    @ReactProp(name = "corners", defaultInt = 0)
    public void setBoxCornerRadius(VGSTextInputLayout view, int radius) {
        view.setBoxCornerRadius(radius, radius, radius, radius);
    }

    public String getFieldName() {
        return input == null ? null : input.getFieldName();
    }

    protected int toDp(Context context, int px) {
        return ResourceUtil.convertPxToDp(context, px);
    }

    private VGSTextInputLayout createLayout(ThemedReactContext reactContext) {
        VGSTextInputLayout layout = new VGSTextInputLayout(reactContext);
        layout.setBoxCornerRadius(DEFAULT_CORNER_RADIUS, DEFAULT_CORNER_RADIUS, DEFAULT_CORNER_RADIUS, DEFAULT_CORNER_RADIUS);
        layout.setHint(getHint());
        return layout;
    }
}

File: android/app/src/main/kotlin/com/collectrndemo/simple/modules/collect/field/core/BaseCollectView.kt
package com.collectrndemo.simple.modules.collect.field.core

import android.R.attr.radius
import android.content.res.ColorStateList
import android.graphics.Color
import com.collectrndemo.simple.modules.OnCreateViewInstanceListener
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.uimanager.ThemedReactContext
import com.facebook.react.uimanager.ViewGroupManager
import com.facebook.react.uimanager.annotations.ReactProp
import com.verygoodsecurity.vgscollect.view.InputFieldView
import com.verygoodsecurity.vgscollect.widget.VGSTextInputLayout


private const val DEFAULT_CORNER_RADIUS = 8

/**
 * @noinspection unused
 */
abstract class BaseCollectView<T : InputFieldView>(
    private val listener: OnCreateViewInstanceListener<InputFieldView>
) : ViewGroupManager<VGSTextInputLayout?>() {

    private lateinit var layout: VGSTextInputLayout

    protected abstract val hint: String?

    protected abstract fun createInput(reactContext: ThemedReactContext?): T

    override fun createViewInstance(reactContext: ThemedReactContext): VGSTextInputLayout {
        layout = createLayout(reactContext)
        val input = createInput(reactContext)
        layout.addView(input)
        listener.onCreateViewInstance(input)
        return layout
    }

    @ReactProp(name = "borderColor")
    fun setBorderColor(view: VGSTextInputLayout, value: String) {
        layout.setBoxStrokeColor(Color.parseColor(value))
        view.invalidate()
    }

    @ReactProp(name = "borderColors")
    fun setBorderColors(view: VGSTextInputLayout, value: ReadableArray) {
        val states = arrayOf(
            intArrayOf(android.R.attr.state_focused), // focused
            intArrayOf(), // unfocused
        )
        val colors = intArrayOf(
            Color.parseColor(value.getString(0)),
            Color.parseColor(value.getString(1)),
        )
        layout.setBoxStrokeColorStateList(ColorStateList(states, colors))
        view.invalidate()
    }

    private fun createLayout(reactContext: ThemedReactContext): VGSTextInputLayout {
        val layout = VGSTextInputLayout(reactContext)
        layout.setBoxCornerRadius(
            DEFAULT_CORNER_RADIUS.toFloat(),
            DEFAULT_CORNER_RADIUS.toFloat(),
            DEFAULT_CORNER_RADIUS.toFloat(),
            DEFAULT_CORNER_RADIUS.toFloat()
        )
        layout.setHintEnabled(false)
        return layout
    }
}
